docs/DesignDocs/RemoteAttestationCollaterals.md

```diff
@@ -0,0 +1,292 @@
+    ///<  |-----------------------|
+    ///<  |  num_elements         |
+    ///<  |-----------------------|
+    ///<  |  buffer_size          |
+    ///<  |-----------------------|
+    ///<  |                       |
+    ///<  |  array(uint32_t)      |
+    ///<  |  size of num_elements |
+    ///<  |-----------------------|
+    ///<  |  Data                 |
+    ///<  |_______________________|
+    ///<
+    uint8_t buffer[];              ///< Buffer of offsets + data
+
+} oe_collaterals_t;
+```
+
+### New Collateral functions
+
+`collaterals.c`
+```C
+/*! \struct sgx_collaterals
+ *
+ * \brief SGX collateral structure
+ *
+ * The generic oe_collaterals_t structure is parsed and converted into this
+ * internal structure.  The order of the generic data elements should
+ * coincide with the order of the fields in this structure.
+ */
+typedef struct _sgx_collaterals_t
+{
+    uint8_t* tcb_info;                      ///< TCB info
+    size_t tcb_info_size;                   ///< TCB Info size
+    uint8_t* tcb_issuer_chain;              ///< PEM format
+    size_t tcb_issuer_chain_size;           ///< Size of the tcb_issuer_chain
+    uint8_t* crl[3];                        ///< CRLs
+    size_t crl_size[3];                     ///< CRLs sizes
+    uint8_t* crl_issuer_chain[3];           ///< PEM format
+    size_t crl_issuer_chain_size[3];        ///< Size of each crl_issuer_chain
+
+    uint8_t* qe_id_info;                    ///< QE Identity info
+    size_t qe_id_info_size;                 ///< QE Identity size
+    uint8_t* qe_id_issuer_chain;            ///< PEM format
+    size_t qe_id_issuer_chain_size;         ///< Size of qe_id_issuer_chain
+
+    uint8_t* creation_datetime;             ///< Time the collaterals were generated
+    size_t create_datetime_size;            ///< The size of creation_datetime.  Should be 24.
+
+} sgx_collaterals_t;
+
+/**
+ * Returns the collaterals needed for verification of a remote OE report.
+ *
+ * This function is only available inside an enclave.
+ *
+ * @param[in] enclave The instance of the enclave that will be used.
+ * @param[in] collaterals_buffer The buffer containing the collaterals to parse.
+ * @param[in] collaterals_buffer_size The size of the **collaterals_buffer**.
+ *
+ * @retval OE_OK The collaterals were successfully retrieved.
+ */
+oe_result_t oe_get_collaterals(
+    uint8_t** collaterals_buffer,
+    size_t* collaterals_buffer_size);
+
+/**
+ * Free up any resources allocated by oe_get_collateras()
+ *
+ * @param[in] collaterals_buffer The buffer containing the collaterals.
+ */
+void oe_free_collaterals(uint8_t* collaterals_buffer);
+
+/**
+ * Verify the integrity of the report and its signature,
+ * with optional collaterals that is associated with the report. If
+ * the collaterals are not specified, this function will fetch
+ * the collaterals.  This only applies to remote reports.  For
+ * local reports please use oe_verify_report().
```

Comment:
Why?   I find this unnecessarily complicated to force a difference on the app developer.
Can't we just have one combined API?   Currently this comes across as over-indexing on an Intel SGX-only design, rather than proposing a processor-agnostic API.

Reply:
The alternative is to update the current function oe_verify_remote_report() to take collaterals as input parameters.

Reply:
I like that idea, as long as those parameters are optional

Reply:
Thanks.

Reply:
My mistake. oe_verify_remote_report() is part of the recently added host_verify library.  A standalone library that can be used to verify OE reports.  The plan is to update this function regardless.  For the OE SDK library, users will need to use oe_verify_report_with_endorsements().  Updating oe_verify_report() is not an option since it will break existing users.

Reply:
even better in my view would be to deprecate (not remove) oe_verify_report and add a new function named something like oe_verify_evidence(), i.e., with that name instead of the name oe_verify_report_with_endorsements, since "report" is an Intel term not an IETF term so would be better to not appear in the API name or structure names going forward.

Reply:
@dthaler
oe_verify_evidence() is a great idea.

Using standard naming is easier to explain their functionalities.